mindtrace/core/mindtrace/core/logging/logger.py

```
@@ -1,6 +1,116 @@
+    name: str = "mindtrace",
+    log_dir: Optional[Path] = None,
+    logger_level: int = logging.DEBUG,
+    stream_level: int = logging.ERROR,
+    file_level: int = logging.DEBUG,
+    file_mode: str = "a",
+    propagate: bool = False,
+    max_bytes: int = 10 * 1024 * 1024,  # 10 MB
+    backup_count: int = 5
+) -> Logger:
+    """
+    Configure and initialize logging for Mindtrace components programmatically.
+
+    Sets up a rotating file handler and a console handler on the given logger.
+    Log file defaults to ~/.cache/mindtrace/{name}.log.
+
+    Args:
+        name (str): Logger name, defaults to "mindtrace".
+        log_dir (Optional[Path]): Custom directory for log file.
+        logger_level (int): Overall logger level.
+        stream_level (int): StreamHandler level (e.g., ERROR).
+        file_level (int): FileHandler level (e.g., DEBUG).
+        file_mode (str): Mode for file handler, default is 'a' (append).
+        propagate (bool): Whether the logger should propagate messages to ancestor loggers.
+        max_bytes (int): Maximum size in bytes before rotating log file.
+        backup_count (int): Number of backup files to retain.
+
+    Returns:
+        Logger: Configured logger instance.
+    """
+    logger = logging.getLogger(name)
+    logger.handlers.clear()
+    logger.setLevel(logger_level)
+    logger.propagate = propagate
+
+    # Set up stream handler
+    stream_handler = logging.StreamHandler()
+    stream_handler.setLevel(stream_level)
+    stream_handler.setFormatter(default_formatter())
+    logger.addHandler(stream_handler)
+
+    # Set up file handler
+    default_config = Config()
+    if name == "mindtrace":
+        child_log_path = f"{name}.log"
+    else:
+        child_log_path = os.path.join("misc", f"{name}.log")
+        
+    if log_dir:
+        log_file_path = os.path.join(log_dir,child_log_path)
+    else:
+        log_file_path = os.path.join(default_config.get('LOGGER').get('LOG_DIR'), child_log_path)
```

I suppose this behavior depends on how the config ends up looking, but I've been assuming it looks like a dict that's only one level deep, in which case the above will throw an error eventually.

In either case, `os.path.join(None, ...)` will also throw an error, which would likely be more obvious to throw at the config/dict part instead. I.e. instead of `get` just do `default_config["LOGGER"]["LOG_DIR"]` which will throw a more obvious error when the expected keys aren't in the config.